google/cloud/bigtable/batcher.py

```diff
@@ -125,22 +298,120 @@ def mutate_rows(self, rows):
+                   mutations.
         """
         for row in rows:
             self.mutate(row)
 
     def flush(self):
-        """Sends the current. batch to Cloud Bigtable.
+        """Sends the current batch to Cloud Bigtable synchronously.
         For example:
 
         .. literalinclude:: snippets.py
             :start-after: [START bigtable_api_batcher_flush]
             :end-before: [END bigtable_api_batcher_flush]
             :dedent: 4
 
+        raises:
+            * :exc:`.batcherMutationsBatchError` if there's any error in the
+                mutations.
+        """
+        rows_to_flush = []
+        while not self._rows.empty():
+            rows_to_flush.append(self._rows.get())
+        response = self.flush_rows(rows_to_flush)
+        return response
+
+    def flush_async(self):
+        """Sends the current batch to Cloud Bigtable asynchronously.
+
+        raises:
+            * :exc:`.batcherMutationsBatchError` if there's any error in the
+                mutations.
+        """
+
+        rows_to_flush = []
+        mutations_count = 0
+        mutations_size = 0
+        rows_count = 0
+        batch_info = BatchInfo()
+
+        while not self._rows.empty():
+            row = self._rows.get()
+            mutations_count += len(row._get_mutations())
+            mutations_size += row.get_mutations_size()
+            rows_count += 1
+            rows_to_flush.append(row)
+            batch_info.mutations_count = mutations_count
+            batch_info.rows_count = rows_count
+            batch_info.mutations_size = mutations_size
+
+            if (
+                rows_count >= self.flush_count
+                or mutations_size >= self.max_row_bytes
+                or mutations_count >= self.flow_control.max_mutations
+                or mutations_size >= self.flow_control.max_row_bytes
+                or self._rows.empty()  # submit when it reached the end of the queue
+            ):
+                self.flow_control.control_flow(batch_info)
+                future = self._executor.submit(self.flush_rows, rows_to_flush)
+                self.futures_mapping[future] = batch_info
+                future.add_done_callback(self.batch_completed_callback)
+
+                # reset and start a new batch
+                rows_to_flush = []
+                mutations_size = 0
+                rows_count = 0
+                mutations_count = 0
+                batch_info = BatchInfo()
+
+    def batch_completed_callback(self, future):
+        """Callback for when the mutation has finished.
+
+        Raise exceptions if there's any.
+        Release the resources locked by the flow control and allow enqueued tasks to be run.
+        """
+
+        processed_rows = self.futures_mapping[future]
+        self.flow_control.release(processed_rows)
+        del self.futures_mapping[future]
+
+    def flush_rows(self, rows_to_flush=None):
```

Comment:
Having both `flush()` and `flush_rows()` seems a little confusing. Especially since this doesn't seem to be "flushing" from the cache in any way. I could see people calling this with no arguments thinking it is the main flush function

Maybe this should be called `mutate_rows`? Or just made into an internal helper function?

Reply:
Agree this is confusing. There is already a public `mutate_rows` from before, and it has different behavior than this one. I'm changing it to private since user isn't expected to call it manually.